utils/src/main/java/software/amazon/awssdk/utils/cache/lru/LruCache.java

```diff
@@ -0,0 +1,223 @@
+/*
+ * Copyright Amazon.com, Inc. or its affiliates. All Rights Reserved.
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License").
+ * You may not use this file except in compliance with the License.
+ * A copy of the License is located at
+ *
+ *  http://aws.amazon.com/apache2.0
+ *
+ * or in the "license" file accompanying this file. This file is distributed
+ * on an "AS IS" BASIS, WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either
+ * express or implied. See the License for the specific language governing
+ * permissions and limitations under the License.
+ */
+
+package software.amazon.awssdk.utils.cache.lru;
+
+import java.util.Map;
+import java.util.Objects;
+import java.util.concurrent.ConcurrentHashMap;
+import java.util.function.Function;
+import software.amazon.awssdk.annotations.SdkProtectedApi;
+import software.amazon.awssdk.utils.Validate;
+
+/**
+ * An LRU (Least Recently Used) cache implementation.
+ * <p>
+ * The user can configure the maximum size of the cache, which is set to a default of 100. Null values are accepted.
+ */
+@SdkProtectedApi
+public final class LruCache<K, V>  {
+
+    private static final int DEFAULT_SIZE = 100;
+
+    private final Map<K, CacheEntry<K, V>> cache;
+    private final Function<K, V> valueSupplier;
+
+    private CacheEntry<K, V> leastRecentlyUsed = null;
+    private CacheEntry<K, V> mostRecentlyUsed = null;
+    private final Object listLock = new Object();
+    private final int maxCacheSize;
+
+    private LruCache(Builder<K, V> b) {
+        this.valueSupplier = b.supplier;
+        Integer customSize = Validate.isPositiveOrNull(b.maxSize, "size");
+        this.maxCacheSize = customSize != null ? customSize : DEFAULT_SIZE;
+        this.cache = new ConcurrentHashMap<>();
+    }
+
+
+    /**
+     * Get a value based on the key. If the value exists in the cache, it's returned, and it's position in the cache is updated.
+     * Otherwise, the value is calculated based on the supplied function {@link Builder#builder(Function)}.
+     */
+    public V get(K key) {
+        CacheEntry<K, V> cachedEntry = cache.computeIfAbsent(key, this::newEntry);
+        updateCacheOrder(cachedEntry);
```

Comment:
Because the cache and list are synchronized independently, it's possible that `cachedEntry` could be evicted when we call `updateCacheOrder`.

We could fix this by adding an 'evicted' flag to each entry that we set to true before we release the list lock during entry eviction, and check is false after we acquire the list lock. Since it's highly unlikely for the 'evicted' flag to be true, if it does happen we could probably just release the lock and redo the get() call.

The alternative to all of this is to just synchronize the whole thing and switch to a non-concurrent map. That might slow down the read path a bit, even if we switch to read/write locks.

Reply:
Evicted flag included